### PR TITLE
Fix for deserialization issue with DS, IS, SV and UV vr types

### LINKED ISSUE #1451: DicomJsonConverter has different behavior than DicomReader/Writer
**Describe the bug**
`DicomReader/Writer` (for converting between raw bytes and in-memory representation) is quite lenient in the validations applied, both on reading and writing. `DicomJsonConverter` applies a number of validations in both the reading and writing process, leading to a behavior gap between the two conversion paths.

For example, a Dicom file could contain an element with Decimal String VR, but invalid content such as `asdf`.

With `DicomReader/Writer`:
- When reading, if `autovalidate` is set to false, the value `asdf` will be used to create a new `DicomDecimalString` using the `IByteBuffer` [constructor](https://github.com/fo-dicom/fo-dicom/blob/b43c485d942d60eb633734764561a83b82f511da/FO-DICOM.Core/DicomElement.cs#L813)
- While in-memory, the value can be retrieved as a `decimal` (which will fail) or as a `string` (which will succeed)
- When writing using `DicomWriter` the value will be written with the same bytes that were read

With `DicomJsonConverter`:
- When reading from JSON, the value `asdf` will not parse as a `decimal` and [fail validation](https://github.com/fo-dicom/fo-dicom/blob/b43c485d942d60eb633734764561a83b82f511da/FO-DICOM.Core/Serialization/JsonDicomConverter.cs#L874). ❌
- When writing to JSON, the value `asdf` (which will be present if the `DicomDataset` was created using `DicomReader`) can be written as a string when the `NumberSerializationMode` is set to `PreferablyAsNumber` or `AsString`.

Ideally, I'd like to see the `DicomJsonConverter` apply the same minimal level of validation as `DicomReader/Writer` during the serialization / deserialization process.

**To Reproduce**
- Read a Dicom file with an element with Decimal String VR and value `asdf` into a `DicomDataset` using `DicomReader`
- Write the `DicomDataset` to JSON using `DicomJsonConverter`
- Read the JSON back into a `DicomDataset` using `DicomJsonConverter`
- An exception is thrown

**Expected behavior**
The content of the dataset is written successfully in DICOM JSON format.

**Environment**
Fellow Oak DICOM version: 5.0.4
OS: Windows 11 x64
Platform: .NET 6

### PR DESCRIPTION
**Pull request by shivakumar-ms**
Fixes [1451](https://github.com/fo-dicom/fo-dicom/issues/1451) “DicomJsonConverter has different behavior than DicomReader/Writer”

As described in [1451](https://github.com/fo-dicom/fo-dicom/issues/1451), reading tags with VR type DS, IS, SV and UV and with a non-numeric string value, from a DicomDataset in JSON string format throws “Malformed DICOM json, number expected“ error. This PR updates the JsonDicomConverter.ReadJsonMultiNumberOrStringValue method to allow the non-numeric value being read as is without any validation if “autoValidate=false”. The values will be read as an IByteBuffer, and use the DicomElement constructor overload the same type. 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [ ] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:

- Created a new abstract generic type DicomMultiStringOrNumberElement to generalize the multi-string number/string element behavior already present in DicomDecimalString and DicomIntegerString. 
- Updated DicomDecimalString, DicomIntegerString, DicomSignedVeryLong, and DicomUnsignedVeryLong types to inherit from this new base class.
- Added a condition to check if the json token type is String, and if autoValidate is set to false. If the condition is true, then add the string value to the list, and mark it as there is a non-numeric string in the list.
- Once all the tokens are read, check if there is a non-numeric string in the list. If true, convert the list into a single string with “\” separator and return it as an IByteBuffer. Otherwise, return the list of values with the specific target type. 

**Testing** 

Added the following unit tests in JsonDicomCoreConverterTest:

- DeserializeDSWithNonNumericValueAsStringDoesNotThrowException 
- DeserializeISWithNonNumericValueAsStringDoesNotThrowException 
- DeserializeSVWithNonNumericValueAsStringDoesNotThrowException 
- DeserializeUVWithNonNumericValueAsStringDoesNotThrowException

### CODE REVIEW

File: FO-DICOM.Core/DicomElement.cs [hunk 1]
```diff
@@ -1599,17 +1599,28 @@ public DicomShortText(DicomTag tag, Encoding[] encodings, IByteBuffer data)
...
+        }
+
+        public DicomSignedVeryLongString(DicomTag tag, params string[] values)
             : base(tag, values)
         {
         }
 
-        public DicomSignedVeryLong(DicomTag tag, IByteBuffer data)
-            : base(tag, data)
+        public DicomSignedVeryLongString(DicomTag tag, params long[] values)
```
Comment (mrbean-bremen) [inline FO-DICOM.Core/DicomElement.cs]: I'm not sure I understand this. `UV` and `SV` are numeric types, not string types, why handle them as strings? I would rather handle the edge case of illegal values in the tag as an exception, not a s a rule, but maybe I don't understand the reasoning here.

File: Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs [hunk 2]
```diff
@@ -104,6 +115,98 @@ public void DeserializeDSAsString()
             Assert.Equal(174.5m, dataset.GetSingleValue<decimal>(DicomTag.PatientSize));
         }
 
+        [Fact]
+        public void DeserializeDSWithInvalidValueAsStringThrowsException()
+        {
+            // in DICOM Standard PS3.18 F.2.3.1 now VRs DS, IS SV and UV may be either number or string
+            var json = @"
```
Comment (mrbean-bremen) [inline Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs]: Ok, I kind of see now what you mean. But this part of the standard is about json types, not DICOM types, it just means that these types can be represented as string or number in the json serialization. In my opinion, this has to be fixed in the json serialization code, not in the DICOM types themselves.

File: FO-DICOM.Core/DicomElement.cs (hunk 1 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/DicomElement.cs]: Since DS, IS, SV and UV VR types allow "String" values, some consumers have "non-numeric" values in these VR type fields. Serialize (write) allows these non-numeric strings; But Deserialize fails. Hence the fix to handle it.

File: Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs (hunk 2 above)
Comment (shivakumar-ms) [reply Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs]: Json serialization using JsonDicomConverter (WriteJson...) allows "non-numeric" values to be serialized without any errors. It fails when we deserialize the resulting json using the same JsonDicomConverter (ReadJson...). Hence this fix.

File: FO-DICOM.Core/DicomElement.cs [hunk 3]
```diff
@@ -789,19 +791,15 @@ private static string[] PrivateDateFormats
...
-        private decimal[] _values;
-
-        #endregion
-
         #region Public Constructors
 
         public DicomDecimalString(DicomTag tag, params decimal[] values)
-            : base(tag, values.Select(x => ToDecimalString(x)).ToArray())
+            : base(tag,
+                  new Func<decimal, string>(x => ToDecimalString(x)),
```
Comment (amoerie) [inline FO-DICOM.Core/DicomElement.cs]: Wouldn't it be better if these conversion functions were private static methods? Then we're not creating lambdas for each instance. If Fellow Oak DICOM targeted .NET 6, we could use static lambdas, but those are not available yet AFAIK on .NET Standard 2.0

File: FO-DICOM.Core/DicomElement.cs [hunk 4]
```diff
@@ -825,55 +823,6 @@ public DicomDecimalString(DicomTag tag, IByteBuffer data)
...
-
-            if (_values == null)
-            {
-                _values =
-                    base.Get<string[]>()
-                        .Select(x => decimal.Parse(x, NumberStyles.Any, CultureInfo.InvariantCulture))
-                        .ToArray();
-            }
-
-            if (typeof(T).GetTypeInfo().IsArray)
```
Comment (amoerie) [inline FO-DICOM.Core/DicomElement.cs:841]: Did you double check that these Get() methods were all the same before you moved them into one implementation. From a first glance, they do look identical.

File: FO-DICOM.Core/DicomElement.cs [hunk 5]
```diff
@@ -1974,4 +1856,103 @@ public DicomUnsignedVeryLong(DicomTag tag, IByteBuffer data)
...
+
+        private TType[] _values;
+
+        private readonly Func<string, TType> _toNumber;
+
+        #endregion
+
+        #region Public Constructors
+
+        public DicomMultiStringOrNumberElement(DicomTag tag, Func<TType, string> toString, Func<string, TType> toNumber, params TType[] values)
```
Comment (amoerie) [inline FO-DICOM.Core/DicomElement.cs:1898]: Is there a specific reason why we don't assign the _values field here with the values constructor parameter?
It seems a bit wasteful to convert the values to strings here, only to re-convert them back to TType when the first call to Get() happens. I assume the reason would be memory related?

File: FO-DICOM.Core/Serialization/JsonDicomConverter.cs [hunk 6]
```diff
@@ -861,7 +862,7 @@ private object ReadJsonMultiNumberOrString<T>(ref Utf8JsonReader reader, GetValu
             }
         }
 
-        private static T[] ReadJsonMultiNumberOrStringValue<T>(ref Utf8JsonReader reader, GetValue<T> getValue, TryParse<T> tryParse)
+        private object ReadJsonMultiNumberOrStringValue<T>(ref Utf8JsonReader reader, GetValue<T> getValue, TryParse<T> tryParse)
```
Comment (amoerie) [inline FO-DICOM.Core/Serialization/JsonDicomConverter.cs:867]: What's the specific reason this method cannot be static anymore? I might be missing something.

File: FO-DICOM.Core/Serialization/JsonDicomConverter.cs [hunk 7]
```diff
@@ -893,8 +900,13 @@ private static T[] ReadJsonMultiNumberOrStringValue<T>(ref Utf8JsonReader reader
...
 
-            var data = childValues.ToArray();
-            return data;
+            if (hasNonNumericString)
+            {
+                var valArray = childValues.Select(x => x.ToString()).ToArray();
+                return ByteConverter.ToByteBuffer(string.Join("\\", valArray));
+            }
+
+            return childValues.Select(x => x).Cast<T>().ToArray();
```
Comment (amoerie) [inline FO-DICOM.Core/Serialization/JsonDicomConverter.cs]: Redundant Select

File: Tests/FO-DICOM.Tests/Serialization/JsonDicomCoreConverterTest.cs [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 using FellowOakDicom.IO.Buffer;
 using FellowOakDicom.Serialization;
+using Newtonsoft.Json;
```
Comment (amoerie) [inline Tests/FO-DICOM.Tests/Serialization/JsonDicomCoreConverterTest.cs]: Why did you need to add this namespace? Because you now also have to fully specify System.Text.Json in several places. Maybe it would be cleaner to only explicitly specify Newtonsoft.Json instead?

**Comment by jovinson-ms** [comment]
Linking #1451 which is a fuller explanation of the issue this PR would like to address.

File: FO-DICOM.Core/DicomElement.cs (hunk 4 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/DicomElement.cs:841]: Yes.

File: FO-DICOM.Core/DicomElement.cs (hunk 5 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/DicomElement.cs:1898]: One reason is we don't want to parse the byte buffer until it is absolutely necessary.

File: FO-DICOM.Core/Serialization/JsonDicomConverter.cs (hunk 7 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/Serialization/JsonDicomConverter.cs]: Fixed.

File: FO-DICOM.Core/Serialization/JsonDicomConverter.cs (hunk 6 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/Serialization/JsonDicomConverter.cs:867]: This method now references the class level member "_autoValidate"

File: FO-DICOM.Core/DicomElement.cs (hunk 3 above)
Comment (shivakumar-ms) [reply FO-DICOM.Core/DicomElement.cs]: Done. These are static readonly now.

File: Tests/FO-DICOM.Tests/Serialization/JsonDicomCoreConverterTest.cs (hunk 8 above)
Comment (shivakumar-ms) [reply Tests/FO-DICOM.Tests/Serialization/JsonDicomCoreConverterTest.cs]: Cleaned up

**Comment by amoerie** [comment]
> Linking #1451 which is a fuller explanation of the issue this PR would like to address.

Does this PR cover the entirety of #1451?

**Review by mrbean-bremen — APPROVED** [review]
Looks good to me now, thanks!

**Review by amoerie — APPROVED** [review]

File: Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs (hunk 2 above)
Comment (gofal) [reply Tests/FO-DICOM.Tests/Serialization/JsonDicomConverterTest.cs]: Then we should fix, that the JsonDicomConverter (WriteJson) will fail on serializing "non-numeric" values.
There is some architectural problem, when json type conversion is added into the code that should deal with DICOM Types.

This issue/pr is about json serialization and deserialization. If this results in resturcturation of DicomElements then separation of concerns is broken.

**Comment by shivakumar-ms** [comment]
> > Linking #1451 which is a fuller explanation of the issue this PR would like to address.
> 
> Does this PR cover the entirety of #1451?

Yes

**Comment by gofal** [comment]
did you take a look at the failing tests?

**Comment by jovinson-ms** [comment]
Can one of the maintainers trigger the tests? @gofal

**Comment by gofal** [comment]
tests are now triggered

**Comment by shivakumar-ms** [comment]
> did you take a look at the failing tests?

@gofal yes, tests are fixed. verified in my local. could you please kick off a test run?

**Comment by shivakumar-ms** [comment]
@gofal @mrbean-bremen @amoerie  All the tests are passing now. could you please approve/merge this PR?

**Comment by amoerie** [comment]
> @gofal @mrbean-bremen @amoerie  All the tests are passing now. could you please approve/merge this PR?

I was on the fence whether to wait for a review from @gofal, but since he already commented here and mentioned nothing aside from the tests, I merged it. Thanks for contributing! 👍

**Comment by shivakumar-ms** [comment]
> > @gofal @mrbean-bremen @amoerie  All the tests are passing now. could you please approve/merge this PR?
> 
> I was on the fence whether to wait for a review from @gofal, but since he already commented here and mentioned nothing aside from the tests, I merged it. Thanks for contributing! 👍

Thanks!!!

**Comment by gofal** [comment]
@amoerie Sadly, you have been too impatient. 
I already did a review and there is this very important concern, that is still unresolved: 
> Then we should fix, that the JsonDicomConverter (WriteJson) will fail on serializing "non-numeric" values.
There is some architectural problem, when json type conversion is added into the code that should deal with DICOM Types.
> This issue/pr is about json serialization and deserialization. If this results in resturcturation of DicomElements then separation of concerns is broken.

I have not done an other review until the unittests have passed, because as long as unit tests failed the PR is still work in progress and many things may still change. 

Futhermore, I guess you have read the discussion in the associated issue, there I described my concernes about this PR very clearly.

**Comment by amoerie** [comment]
> @amoerie Sadly, you have been too impatient. 
> I already did a review and there is this very important concern, that is still unresolved: 
> > Then we should fix, that the JsonDicomConverter (WriteJson) will fail on serializing "non-numeric" values.
> There is some architectural problem, when json type conversion is added into the code that should deal with DICOM Types.
> > This issue/pr is about json serialization and deserialization. If this results in resturcturation of DicomElements then separation of concerns is broken.
> 
> I have not done an other review until the unittests have passed, because as long as unit tests failed the PR is still work in progress and many things may still change. 
> 
> Futhermore, I guess you have read the discussion in the associated issue, there I described my concernes about this PR very clearly.

Oh my apologies! Perhaps we can simply revert the PR for now?
